spyder/api/widgets.py

```diff
@@ -0,0 +1,1478 @@
+
+    def add_action(self, action, section=None, before=None):
+        """
+        Add action to a given menu section.
+        """
+        if before is None:
+            self._actions.append((section, action))
+        else:
+            new_actions = []
+            for sec, act in self._actions:
+                if act == before:
+                    new_actions.append((section, action))
+
+                new_actions.append((sec, act))
+
+            self._actions = new_actions
+
+        if section not in self._sections:
+            self._sections.append(section)
+
+        # Track state of menu to avoid re-rendering if menu has not changed
+        self._dirty = True
+        self._ordered_actions = []
+
+    def get_name(self):
+        """
+        Return name for menu.
+        """
+        return self._name
+
+    def get_sections(self):
+        """
+        Return a tuple of menu sections.
+        """
+        return tuple(self._sections)
+
+    def _render(self):
+        """
+        Create the menu prior to showing it. This takes into account sections
+        and location of menus. It also hides consecutive separators if found.
+        """
+        if self._dirty:
+            self.clear()
+            actions = []
+            for section in self._sections:
+                for (sec, action) in self._actions:
+                    if sec == section:
+                        actions.append(action)
+
+                actions.append(None)
+
+            # TODO: Remove any consecutive Nones
+            add_actions(self, actions)
+
+            self._ordered_actions = actions
+            self._dirty = False
+
+    @property
+    def name(self):
+        """
+        Return name for menu.
+        """
+        return self._name
+
+    @property
+    def sections(self):
+        """
+        Return a tuple of menu sections.
+        """
+        return tuple(self._sections)
+
+
+class MainWidgetMenu(SpyderMenu):
+    """
+    This menu fixes the bottom section of the options menu.
+    """
+
+    def _render(self):
+        """
+        Create the menu prior to showing it. This takes into account sections
+        and location of menus. It also hides consecutive separators if found.
+        """
+        if self._dirty:
+            self.clear()
+            bottom = 'bottom'
+            actions = []
+            for section in self._sections:
+                for (sec, action) in self._actions:
+                    if sec == section and sec != bottom:
+                        actions.append(action)
+
+                actions.append(None)
+
+            # Add bottom actions
+            for (sec, action) in self._actions:
+                if sec == bottom:
+                    actions.append(action)
+
+            # TODO: Remove any consecutive Nones
+            add_actions(self, actions)
+
+            self._ordered_actions = actions
+            self._dirty = False
+
+
+class ApplicationMenu(SpyderMenu):
```

Comment:
I don't understand the purpose of this empty class.

Reply:
Ensure that the methods will receive the appropriate Menu. This menu will need to perform some extra checks. I will add this as soon as I start working with a plugin that creates a lot of actions inside an app menu (like editor or console)